src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -139,24 +139,27 @@ private static IntPtr HostFxrResolver(Assembly assembly, string libraryName)
                 return IntPtr.Zero;
             }
 
-            string hostFxrRoot = Path.Combine(s_dotnetPath.Value, "host", "fxr");
-            if (Directory.Exists(hostFxrRoot))
+            foreach (string dotnetPath in s_dotnetPathCandidates.Value)
             {
-                // Load hostfxr from the highest version, because it should be backward-compatible
-                SemanticVersion? hostFxrAssemblyDirectory = Directory.GetDirectories(hostFxrRoot)
-                            .Max(str => SemanticVersionParser.TryParse(str, out SemanticVersion? version) ? version : null);
-
-                if (hostFxrAssemblyDirectory != null && !string.IsNullOrEmpty(hostFxrAssemblyDirectory.OriginalValue))
+                string hostFxrRoot = Path.Combine(dotnetPath, "host", "fxr");
+                if (Directory.Exists(hostFxrRoot))
                 {
-                    string hostFxrAssembly = Path.Combine(hostFxrAssemblyDirectory.OriginalValue, Path.ChangeExtension(hostFxrLibName, libExtension));
+                    // Load hostfxr from the highest version, because it should be backward-compatible
+                    SemanticVersion? hostFxrAssemblyDirectory = Directory.GetDirectories(hostFxrRoot)
+                                .Max(str => SemanticVersionParser.TryParse(str, out SemanticVersion? version) ? version : null);
 
-                    if (File.Exists(hostFxrAssembly))
+                    if (hostFxrAssemblyDirectory != null && !string.IsNullOrEmpty(hostFxrAssemblyDirectory.OriginalValue))
                     {
-                        return NativeLibrary.TryLoad(hostFxrAssembly, out IntPtr handle) ? handle : IntPtr.Zero;
+                        string hostFxrAssembly = Path.Combine(hostFxrAssemblyDirectory.OriginalValue, Path.ChangeExtension(hostFxrLibName, libExtension));
+
+                        if (File.Exists(hostFxrAssembly))
+                        {
+                            return NativeLibrary.TryLoad(hostFxrAssembly, out IntPtr handle) ? handle : IntPtr.Zero;
```

Comment:
Given the issues with running ad-hoc signed code on Mac, I wonder if we shouldn't make this part also more defensive. If `TryLoad` fails for the highest version, we could try to call one from the second highest, and so on.

Basically (pseudo-code):

```
for all dotnet location candidates:
  for all host\fxr subdirectories ordered from semantically highest version:
    if TryLoad() succeeds, return its result, otherwise continue
```

cc @lifengl

Reply:
done